### PR TITLE
Make all utilities use waitForIdleState

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This makes sure `waitForIdleState()` is the only place where we still do a `Thread.sleep()`. Having this isolated makes it easier to remove it once we've implemented a way to wait for an idle state in the engine.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [ ] Javadoc has been written
* [ ] The documentation is updated

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
@romansmirnov Would you be willing to review this PR, since Peter is on holiday this week? I'd be happy to answer any questions you might have.

**Comment by remcowesterhoud** [comment]
Hey @korthout! Do you have time to have a look at this one? It's a relatively small refactoring.
Some more context:

EZE does not have a way to check if the engine is idle. This is something we will need for this project, but was not our first priority to build. Therefore we started using `Thread.sleep()` in our unit tests. We tried to keep these sleeps as isolated as possible, which makes it easier to remove them later on.
At some point we introduced a `waitForIdleState()` method. This would do the `Thread.sleep()` for now, but the goal is to only have to change this single method to check if the engine is idle to get rid of all these dirty sleeps.

This PR makes sure the last places where `Thread.sleep()` was still being used are replaced with the `waitForIdleState` method.

File: src/test/java/io/camunda/testing/util/Utilities.java [hunk 1]
```diff
@@ -77,13 +76,15 @@ public static DeploymentEvent deployProcesses(
   }
 
   public static ProcessInstanceEvent startProcessInstance(
-      final ZeebeClient client, final String processId) throws InterruptedException {
-    return startProcessInstance(client, processId, new HashMap<>());
+      final ZeebeEngine engine, final ZeebeClient client, final String processId) {
+    return startProcessInstance(engine, client, processId, new HashMap<>());
```
Comment (korthout) [inline src/test/java/io/camunda/testing/util/Utilities.java:80]: This pull request makes considerable changes to the public interface, mainly by adding `engine` as an additional parameter to these methods. That is of course necessary in order to use the `waitForIdleState` method, but with the number of parameters increasing we could consider improving usability, moving common parameters to a constructor and no longer having static methods.

❓ Is this something you have considered? Do we have a goal design in mind for the interfaces? 

I ask this because changing such interfaces after an initial release might be difficult due to breaking changes.

**Review by korthout — COMMENTED** [review]
Please consider my comments, specifically I worry that not all methods can wait for an idle state, but should rather wait for a specific goal to be reached like process instance created or message published. Perhaps my idea of an idle state is not the same as what is meant here. I assume it means, the process instance waits for external input to continue its processing.

File: src/test/java/io/camunda/testing/util/Utilities.java [hunk 2]
```diff
@@ -92,7 +93,7 @@ public static ProcessInstanceEvent startProcessInstance(
             .variables(variables)
             .send()
             .join();
-    Thread.sleep(100);
+    waitForIdleState(engine);
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply src/test/java/io/camunda/testing/util/Utilities.java:96]: Idle state is still a bit vague. We consider the engine to be idle when it is not actively processing. One of the challenges we have is figuring out how to notice this, but that's not relevant to the current PR. It could be as simple as not processing for X amount of time for the MVP.

If a process has no "waiting" elements (timers, user tasks, message events), the engine will complete the entire process. Once the process has been completed the engine does not have any processing to do, thus the engine is considered to be idle.

**Comment by korthout** [comment]
@remcowesterhoud It cost me more than an hour (perhaps an hour and a half) to review this changeset, even though it was supposed to be a relatively small refactoring. The biggest reason for this was that I didn't only had to review the requested change, but also other refactorings which were mixed-in with the main change, which meant I had to search for it.

My tip for a next PR, is to individually commit refactorings that lead to "larger" changesets. For example in this PR, when adding `engine` as the first method parameter of `sendMessage` and `startProcessInstance`, for removing the `throws Exception`, for wrapping lines at a different places, etc. That would've helped me practically skip hundreds of lines of changes.

File: src/test/java/io/camunda/testing/util/Utilities.java [hunk 3]
```diff
@@ -131,17 +134,17 @@ public static PublishMessageResponse sendMessage(
...
     return response;
   }
 
-  public static void increaseTime(final ZeebeEngineClock clock, final Duration duration)
-      throws InterruptedException {
-    clock.increaseTime(duration);
-    Thread.sleep(100);
+  public static void increaseTime(final ZeebeEngine engine, final Duration duration) {
+    engine.clock().increaseTime(duration);
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply src/test/java/io/camunda/testing/util/Utilities.java:142]: I think it does relate. Before we passed the clock which we used for `increaseTime()`. Now that we replaced the `Thread.sleep()` with the `waitForIdleState()` we needed access to the `ZeebeEngine`. This means we can either pass both the `ZeebeClock` and the `ZeebeEngine` as parameters. However, the engine already contains the `ZeebeClock` there is no need to pass them both.

Another thing to note is that this is not a public interface. This is just a utility class we use for our own unit tests. Nothing of this is exposed to the outside world.

File: src/test/java/io/camunda/testing/util/Utilities.java (hunk 1 above)
Comment (remcowesterhoud) [reply src/test/java/io/camunda/testing/util/Utilities.java:80]: As mentioned in the comment above, this is not a public interface. Therefore I don't see a problem in the way it is currently. Eventually we will build a public interface and at that point we will also start thinking it through more thoroughly, and this code will probably be replaced.

This is not something that will be done this quarter though.

File: src/test/java/io/camunda/testing/util/Utilities.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (korthout) [reply src/test/java/io/camunda/testing/util/Utilities.java:96]: It might be an edge case, but I'm trying to imagine processes without waiting elements that don't in 'reasonable' time. Perhaps we don't have to care about it now 👍

File: src/test/java/io/camunda/testing/util/Utilities.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (korthout) [reply src/test/java/io/camunda/testing/util/Utilities.java:142]: >this is not a public interface.

Ah! I was not aware. Now that I see it's in the test sources it makes sense. Never mind my carefulness. 👍

File: src/test/java/io/camunda/testing/util/Utilities.java (hunk 1 above)
Comment (korthout) [reply src/test/java/io/camunda/testing/util/Utilities.java:80]: Never mind my carefulness. 👍

See comment in above conversation.

**Comment by remcowesterhoud** [comment]
@korthout Thanks for the review and the feedback! I'll keep that in mind. Most of this was automatic formatting. For me it seemed clear, but I can image that would be different when you have less context.

I've commented on your review comments. I think the main misconception was that I was editing a public interface. As explained in my comments this is not the case, it's just our own testing utility to prevent duplication. Please read me comments and let me know if it is good to merge!

**Review by korthout — APPROVED** [review]
LGTM 👍
